test/e2e/storage/csi_hostpath.go

```diff
@@ -0,0 +1,199 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// This file is used to deploy the CSI hostPath plugin
+// More Information: https://github.com/kubernetes-csi/drivers/tree/master/pkg/hostpath
+
+package storage
+
+import (
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/test/e2e/framework"
+)
+
+const (
+	csiHostPathPluginImage string = "docker.io/k8scsi/hostpathplugin:0.1"
+)
+
+func csiHostPathPod(
+	client clientset.Interface,
+	config framework.VolumeTestConfig,
+	teardown bool,
+	f *framework.Framework,
+	sa *v1.ServiceAccount,
+) *v1.Pod {
+	podClient := client.CoreV1().Pods(config.Namespace)
+
+	priv := true
+	mountPropagation := v1.MountPropagationBidirectional
+	hostPathType := v1.HostPathDirectoryOrCreate
+	pod := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      config.Prefix + "-pod",
+			Namespace: config.Namespace,
+			Labels: map[string]string{
+				"app": "hostpath-driver",
+			},
+		},
+		Spec: v1.PodSpec{
+			ServiceAccountName: sa.GetName(),
+			NodeName:           config.ServerNodeName,
+			RestartPolicy:      v1.RestartPolicyNever,
+			Containers: []v1.Container{
+				{
+					Name:            "external-provisioner",
+					Image:           csiExternalProvisionerImage,
+					ImagePullPolicy: v1.PullAlways,
+					Args: []string{
+						"--v=5",
+						"--provisioner=csi-hostpath",
+						"--csi-address=/csi/csi.sock",
+					},
+					VolumeMounts: []v1.VolumeMount{
+						{
+							Name:      "socket-dir",
+							MountPath: "/csi",
+						},
+					},
+				},
+				{
+					Name:            "driver-registrar",
```

Comment:
I see that you bundled the all the external components and the driver in to the same pod. Not a huge deal, but our recommendation is two pods: 1) no more then one replica containing external-attacher, external-provisioner, and csi-driver and 2) on every node a driver-registrar and csi-driver.

We can consider this configuration for additional E2E tests in the future.